internal/auth/repository_auth_method.go

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package auth
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/pagination"
+	"github.com/hashicorp/boundary/internal/util"
+	"github.com/hashicorp/go-kms-wrapping/v2/extras/structwrapping"
+)
+
+// AuthMethodRepository coordinates calls across different subtype services
+// to gather information about all auth methods.
+type AuthMethodRepository struct {
+	reader db.Reader
+	writer db.Writer
+	kms    *kms.Kms
+}
+
+// NewAuthMethodRepository returns a new auth method repository.
+func NewAuthMethodRepository(ctx context.Context, reader db.Reader, writer db.Writer, kms *kms.Kms) (*AuthMethodRepository, error) {
+	const op = "auth.NewAuthMethodRepository"
+	switch {
+	case util.IsNil(reader):
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing DB reader")
+	case util.IsNil(writer):
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing DB writer")
+	}
+	return &AuthMethodRepository{
+		reader: reader,
+		writer: writer,
+		kms:    kms,
+	}, nil
+}
+
+// List lists auth methods across all subtypes.
+func (amr *AuthMethodRepository) List(ctx context.Context, scopeIds []string, afterItem pagination.Item, limit int) ([]AuthMethod, time.Time, error) {
+	const op = "auth.(*AuthMethodRepository).list"
+	switch {
+	case len(scopeIds) == 0:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing scope ids")
+	case limit < 1:
+		return nil, time.Time{}, errors.New(ctx, errors.InvalidParameter, op, "missing limit")
+	}
+
+	var inClauses []string
+	var args []any
+	for i, scopeId := range scopeIds {
+		arg := "scope_id_" + strconv.Itoa(i)
+		inClauses = append(inClauses, "@"+arg)
+		args = append(args, sql.Named(arg, scopeId))
+	}
+	inClause := strings.Join(inClauses, ", ")
+	whereClause := "scope_id in (" + inClause + ")"
```

Comment:
Apparently it _is_ possible to do this with
```suggestion
	whereClause := "scope_id in (@scope_ids)"
	args := []any{sql.Named("scope_ids", scopeIds)}
```
. I haven't yet tested it, but it's what Todd did in Roles. I intend to submit a PR to do this for the rest of the places we do this dance so you might want to try it here too.

Reply:
https://github.com/hashicorp/boundary/pull/4162